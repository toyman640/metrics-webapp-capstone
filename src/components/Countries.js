import { useSelector, useDispatch } from 'react-redux';
import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { getCountries } from '../redux/countries/countrySlice';
import Search from './Search';
import africa from '../africa-map.png';

function Countries() {
  const dispatch = useDispatch();
  const countriesArr = useSelector((state) => state.countries.countrydata);
  const status = useSelector((state) => state.countries.status);
  const [filteredCountries, setFilteredCountries] = useState([]);
  const [searchValue, setSearchValue] = useState('');

  useEffect(() => {
    dispatch(getCountries());
  }, [dispatch]);

  useEffect(() => {
    if (countriesArr) {
      setFilteredCountries(
        countriesArr.filter((country) => country.name
          .toLowerCase().includes(searchValue.toLowerCase())),
      );
    }
  }, [countriesArr, searchValue]);

  const handleSearch = (value) => {
    setSearchValue(value);
  };

  if (status === 'idle' || status === 'pending' || !countriesArr) {
    // Handle loading state
    return <div>Loading...</div>;
  }

  return (
    <div className="Main-div">
      <div className="Top-section">
        <img className="Africa-map" src={africa} alt="map of africa" />
        <div className="Second-part">
          <h2>Africa</h2>
          <Search onSearch={handleSearch} />
        </div>
      </div>
      <div className="Countries-grid">
        {searchValue && filteredCountries.length === 0 ? (
          <div>No countries found.</div>
        ) : (
          filteredCountries.map((country) => (
            <div className="Country-list" key={country.id}>
              <Link className="Country-link" to={`/country/${country.id}`} key={country.id}>
                <div
                  className="Flag-div"
                  style={{
                    backgroundImage: `url(${country.image})`,
                    backgroundSize: 'cover',
                    backgroundRepeat: 'no-repeat',
                    position: 'relative',
                  }}
                >
                  <h4 className="Country-name">{country.name}</h4>
                </div>
              </Link>
            </div>
          ))
        )}
      </div>
    </div>
  );
}

export default Countries;
