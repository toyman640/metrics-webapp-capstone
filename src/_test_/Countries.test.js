import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Countries from '../components/Countries';

const mockStore = configureStore([]);

jest.mock('axios', () => ({
  get: jest.fn(() => Promise.resolve({
    data: [
      {
        id: 1,
        name: 'Country 1',
        image: 'image-url-1',
      },
      {
        id: 2,
        name: 'Country 2',
        image: 'image-url-2',
      },
    ],
  })),
}));

describe('Countries', () => {
  let store;
  const initialState = {
    countries: {
      countrydata: [],
      status: 'succeeded',
    },
  };

  beforeEach(() => {
    store = mockStore(initialState);
  });

  test('renders countries from the store', async () => {
    const component = renderer.create(
      <Provider store={store}>
        <Countries />
      </Provider>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
