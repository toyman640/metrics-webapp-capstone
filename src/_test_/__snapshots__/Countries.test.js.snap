// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countries renders countries from the store 1`] = `
<div
  className="Main-div"
>
  <div
    className="Top-section"
  >
    <img
      alt="map of africa"
      className="Africa-map"
      src="africa-map.png"
    />
    <div
      className="Second-part"
    >
      <h2>
        Africa
      </h2>
      <div>
        <form
          className="Search-form"
        >
          <input
            className="Text-box"
            onChange={[Function]}
            placeholder="Click here to search.."
            type="text"
          />
        </form>
      </div>
    </div>
  </div>
  <div
    className="Countries-grid"
  />
</div>
`;
